### PR TITLE
Leverage CMake to automatically pull dependencies at configure time if not present

### PR DESCRIPTION
**Pull request by theblackunknown**

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -1,39 +1,20 @@
 cmake_minimum_required(VERSION 3.0)
 
-option(WARN_VERBOSE "Enable highest warning level and disable warnings as errors" OFF)
+project(shaderunner)
```
Comment (theblackunknown) [inline CMakeLists.txt:3]: I kept a single project instead of multiple ones, feel free to revert if that is not intended

File: CMakeLists.txt [hunk 2]
```diff
@@ -1,39 +1,20 @@
...
-set(INCLUDE_DIR sources/inc)
 set(LIB_DIR lib)
 
 # ==============================================================================
-
-set( LIB_INCLUDE_DIR
-	 ${LIB_DIR}/glew/include/
-	 ${LIB_DIR}/boost/
-	 ${LIB_DIR}/imgui/
-	 )
```
Comment (theblackunknown) [inline CMakeLists.txt:16]: This is now managed through `target_link_libraries` now that linked target declared properly their dependencies

File: CMakeLists.txt [hunk 3]
```diff
@@ -1,39 +1,20 @@
...
-
-function(_FIND_GL_LIBRARIES OUT)
-	find_package(OpenGL REQUIRED)
-	find_library(GLEW_LIB glew32 PATHS ${LIB_DIR}/glew/lib/Release/x64/)
-	set(${OUT} "${GLEW_LIB}" ${OPENGL_gl_LIBRARY} PARENT_SCOPE)
-endfunction()
```
Comment (theblackunknown) [inline CMakeLists.txt:22]: Refactored as simple `find_package` or CMake IMPORTED targets

File: CMakeLists.txt [hunk 4]
```diff
@@ -1,39 +1,20 @@
...
-	 ${IMGUI_DIR}/imgui_draw.cpp
-	 )
-
-project(imgui)
-include_directories(${IMGUI_DIR})
-add_library(imgui STATIC ${IMGUI_SOURCES})
-
+find_package(OpenGL REQUIRED)
+# ==============================================================================
+include(cmake/external/glew.cmake)
```
Comment (theblackunknown) [inline CMakeLists.txt:15]: Externalized in another file to be less verbose

File: CMakeLists.txt [hunk 5]
```diff
@@ -1,39 +1,20 @@
...
-	 ${LIB_DIR}/imgui/
-	 )
-
-function(_FIND_GL_LIBRARIES OUT)
-	find_package(OpenGL REQUIRED)
-	find_library(GLEW_LIB glew32 PATHS ${LIB_DIR}/glew/lib/Release/x64/)
-	set(${OUT} "${GLEW_LIB}" ${OPENGL_gl_LIBRARY} PARENT_SCOPE)
-endfunction()
-
+include(cmake/external/boost.cmake)
```
Comment (theblackunknown) [inline CMakeLists.txt:11]: Externalized in another file to be less verbose

File: CMakeLists.txt [hunk 6]
```diff
@@ -67,86 +48,101 @@ set( WIN32_BOOTSTRAP_SOURCES
...
-	add_compile_options(-std:c++latest)
+function(_COMMON_PROJECT_OPTIONS TARGET_NAME)
+	target_include_directories(${TARGET_NAME}
+		PRIVATE
+			sources/inc
+	)
+	target_link_libraries(${TARGET_NAME}
+		PRIVATE
+			Boost::boost
+			Boost::filesystem
```
Comment (theblackunknown) [inline CMakeLists.txt:59]: Whether it is through `FindBoost` or automatic dependency, you should now used proper CMake targets for this

File: CMakeLists.txt [hunk 7]
```diff
@@ -67,86 +48,101 @@ set( WIN32_BOOTSTRAP_SOURCES
 
 # ==============================================================================
 
-function(_COMMON_PROJECT_OPTIONS)
-	include_directories(${INCLUDE_DIR} ${LIB_INCLUDE_DIR})
-	link_directories(${LIB_DIR}/boost/lib/)
-
-	add_compile_options(-std:c++latest)
+function(_COMMON_PROJECT_OPTIONS TARGET_NAME)
```
Comment (theblackunknown) [inline CMakeLists.txt:51]: I added a target argument to avoid to globally declared include directories or linked libraries

File: CMakeLists.txt [hunk 8]
```diff
@@ -67,86 +48,101 @@ set( WIN32_BOOTSTRAP_SOURCES
...
+	PRIVATE
+		oglbase
+		uibase
+		shaderunner
+)
+_common_project_options(win32_bootstrap)
+
+get_target_property(GLEW_DSO_PATH
+	GLEW::GLEW
+	IMPORTED_LOCATION
```
Comment (theblackunknown) [inline CMakeLists.txt:123]: This is the property pointing to the DLL file
NOTE: If one day you switch it to a static library you may require to change this portion

File: cmake/external/imgui.cmake [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+    FetchContent_GetProperties(imgui)
+    if(NOT imgui_POPULATED)
+        message(STATUS "[imgui] not found on the system, fetching it...")
+      FetchContent_Populate(imgui)
+    endif()
+
+    message(STATUS "[imgui] directory: ${imgui_SOURCE_DIR}")
+
+    add_library(_imgui STATIC)
+    add_library(imgui::imgui ALIAS _imgui)
```
Comment (theblackunknown) [inline cmake/external/imgui.cmake:22]: Declared as a static librayr, as you requested :)

File: CMakeLists.txt [hunk 10]
```diff
@@ -1,39 +1,20 @@
 cmake_minimum_required(VERSION 3.0)
```
Comment (JabberSnatch) [inline CMakeLists.txt]: should have updated the version number, to get a better message than "Missing function or something like that

File: CMakeLists.txt (hunk 8 above)
Comment (JabberSnatch) [reply CMakeLists.txt:123]: Please insert the "NOTE" in the code

**Review by JabberSnatch — CHANGES_REQUESTED** [review]
Trying the system right now, if it works, I'll accept the pull request with the changes I mentionned in the comments.
Thanks for the cmake work

**Comment by JabberSnatch** [comment]
Build worked on the first try, with one warning that I didn't like : CMake did not find a "Find[lib].cmake" file.
Would you consider adding a 'quiet' argument to every find_package call ? The rationale being that these calls are expected to fail on first try.

File: CMakeLists.txt (hunk 10 above)
Comment (theblackunknown) [reply CMakeLists.txt]: Updated

**Comment by theblackunknown** [comment]
I adressed your comments, removed debug status message, everything should works fine now.
Let me know if something else remain to be done

File: CMakeLists.txt (hunk 8 above)
Comment (theblackunknown) [reply CMakeLists.txt:123]: Not sure what you meant by "NOTE" but I added a comment to explain what this call does

**Comment by theblackunknown** [comment]
Just realised that by being twisted by Gerrit, it is not trivial to follow update in Github pull request the way I did a forced update

**Pull request merged**